cvat/apps/dataset_manager/views.py

```diff
@@ -32,78 +36,123 @@ def log_exception(logger=None, exc_info=True):
-                "The {} '{}' is exported as '{}' at '{}' "
-                "and available for downloading for the next {}. "
-                "Export cache cleaning job is enqueued, id '{}'".format(
-                    db_instance.__class__.__name__.lower(),
-                    db_instance.name if isinstance(db_instance, (Project, Task)) else db_instance.id,
-                    dst_format, output_path, cache_ttl,
-                    cleaning_job.id
-                ))
+            tasks_update = list(map(
+                lambda db_task: timezone.localtime(db_task.updated_date),
+                db_instance.tasks.all()
+            ))
+            instance_update_time = max(tasks_update + [instance_update_time])
+
+        output_path = '%s-instance%s_%s.%s' % (
+            'dataset' if save_images else 'annotations',
+            # store the instance timestamp in the file name to reliably get this information
+            # ctime / mtime do not return file creation time on linux
+            # mtime is used for file usage checks
+            instance_update_time.timestamp(),
+            make_file_name(to_snake_case(dst_format)),
+            exporter.EXT,
+        )
+        output_path = osp.join(cache_dir, output_path)
+
+        os.makedirs(cache_dir, exist_ok=True)
+
+        with get_dataset_cache_lock(
+            output_path,
+            block=True,
+            acquire_timeout=EXPORT_CACHE_LOCK_TIMEOUT,
+            ttl=rq.get_current_job().timeout,
+        ):
+            if not (
+                osp.exists(output_path)
+                and instance_update_time.timestamp() <= get_file_instance_timestamp(output_path)
+            ):
+                with tempfile.TemporaryDirectory(dir=cache_dir) as temp_dir:
+                    temp_file = osp.join(temp_dir, 'result')
+                    export_fn(db_instance.id, temp_file, dst_format,
+                        server_url=server_url, save_images=save_images)
+                    os.replace(temp_file, output_path)
+
+                scheduler = django_rq.get_scheduler(settings.CVAT_QUEUES.EXPORT_DATA.value)
+                cleaning_job = scheduler.enqueue_in(
+                    time_delta=cache_ttl,
+                    func=clear_export_cache,
+                    file_path=output_path,
+                    file_ctime=instance_update_time.timestamp(),
+                    logger=logger
+                )
+                logger.info(
+                    "The {} '{}' is exported as '{}' at '{}' "
+                    "and available for downloading for the next {}. "
+                    "Export cache cleaning job is enqueued, id '{}'".format(
+                        db_instance.__class__.__name__.lower(),
+                        db_instance.name if isinstance(db_instance, (Project, Task)) else db_instance.id,
+                        dst_format, output_path, cache_ttl,
+                        cleaning_job.id
+                    )
+                )
```

Comment:
@coderabbitai, yes, it would be nice to get some advice from you. This PR fixes thread/data race condition on cached file removal (which is removed automatically after some delay). Do you have any ideas on how it can be tested reliably?

Reply:
@coderabbitai, added.

Reply:
Could you generate additional tests for checking file removal, file downloading, successful export, retries in export, and retries in file removal?